source/common/tcp/conn_pool.h

```diff
@@ -0,0 +1,203 @@
+
+#include "envoy/event/deferred_deletable.h"
+#include "envoy/event/timer.h"
+#include "envoy/network/connection.h"
+#include "envoy/network/filter.h"
+#include "envoy/stats/timespan.h"
+#include "envoy/tcp/conn_pool.h"
+#include "envoy/upstream/upstream.h"
+
+#include "common/common/linked_object.h"
+#include "common/common/logger.h"
+#include "common/http/conn_pool_base.h"
+#include "common/network/filter_impl.h"
+
+namespace Envoy {
+namespace Tcp {
+
+class ConnPoolImpl;
+
+class TcpPendingRequest : public Envoy::ConnectionPool::PendingRequest {
+public:
+  TcpPendingRequest(Envoy::ConnectionPool::ConnPoolImplBase& parent,
+                    Tcp::ConnectionPool::Callbacks& callbacks)
+      : Envoy::ConnectionPool::PendingRequest(parent), callbacks_(callbacks) {}
+  void* context() override { return static_cast<void*>(&callbacks_); }
+
+  Tcp::ConnectionPool::Callbacks& callbacks_;
+};
+
+class ActiveTcpClient : public Envoy::ConnectionPool::ActiveClient {
+public:
+  struct ConnReadFilter : public Network::ReadFilterBaseImpl {
+    ConnReadFilter(ActiveTcpClient& parent) : parent_(parent) {}
+
+    // Network::ReadFilter
+    Network::FilterStatus onData(Buffer::Instance& data, bool end_stream) override {
+      parent_.onUpstreamData(data, end_stream);
+      return Network::FilterStatus::StopIteration;
+    }
+    ActiveTcpClient& parent_;
+  };
+
+  // This acts as the bridge between the ActiveTcpClient and an individual TCP connection.
+  class TcpConnectionData : public Envoy::Tcp::ConnectionPool::ConnectionData {
+  public:
+    TcpConnectionData(ActiveTcpClient& parent, Network::ClientConnection& connection)
+        : parent_(&parent), connection_(connection) {
+      parent_->tcp_connection_data_ = this;
+    }
+    ~TcpConnectionData() override {
+      // Generally it is the case that TcpConnectionData will be destroyed before the
+      // ActiveTcpClient. Because ordering on the deferred delete list is not guaranteed in the
+      // case of a disconnect, make sure parent_ is valid before doing clean-up.
+      if (parent_) {
+        parent_->clearCallbacks();
+      }
+    }
+
+    Network::ClientConnection& connection() override { return connection_; }
+    void setConnectionState(ConnectionPool::ConnectionStatePtr&& state) override {
+      parent_->connection_state_ = std::move(state);
+    }
+
+    void addUpstreamCallbacks(ConnectionPool::UpstreamCallbacks& callbacks) override {
+      parent_->callbacks_ = &callbacks;
+    }
+    void release() { parent_ = nullptr; }
+
+  protected:
+    ConnectionPool::ConnectionState* connectionState() override {
+      return parent_->connection_state_.get();
+    }
+
+  private:
+    ActiveTcpClient* parent_;
+    Network::ClientConnection& connection_;
+  };
+
+  ActiveTcpClient(ConnPoolImpl& parent, uint64_t lifetime_request_limit,
+                  uint64_t concurrent_request_limit);
+  ~ActiveTcpClient() override;
+
+  void onEvent(Network::ConnectionEvent event) override;
```

Comment:
Please add the standard comment for which interface these are overriding/implementing